### ISSUE TITLE
[Bug]: Impersonation not cleared when map made invisible.

### ISSUE THREAD

**Issue by kwvanderlinde** [RPTools/maptool#4387]
### Describe the Bug

As a player, if I am impersonating a token on a map, and the map is then hidden by the GM, I am sometimes left impersonating the token and able to click macro buttons for it in the _Impersonate_ window. Using such a macro button results in this error:
> Could not execute the command: "im 57B2B61004944219903FAC1794A3B980:[r: "hi"]", exception = Cannot invoke "net.rptools.maptool.client.ui.zone.ZoneRenderer.getZone()" because the return value of "net.rptools.maptool.client.ui.MapToolFrame.getCurrentZoneRenderer()" is null

### To Reproduce

1. Open or create a campaign with a token owned by _All Players_
2. Start a server, and connect to it as a player.
3. As a player, impersonate a token.
4. As a GM, hide the map containing the impersonated token.
5. As a player, observe the _Impersonate_ window still shows the token's macros.

### Expected Behaviour

Hiding a map should consistently clear any player impersonations for tokens on the map. That includes any functionality that relies on impersonation: the _Impersonated_ window, chat window, in macros, etc.

### Screenshots

These are taken from a player window where the map has just been hidden:
![image](https://github.com/RPTools/maptool/assets/7492219/51077654-676a-4ce6-b818-8aca282482bf)
![image](https://github.com/RPTools/maptool/assets/7492219/6bb0bb08-3742-4421-a987-2be5544e8e1e)

Notice also that the top screenshot does not show the impersonated icon as it would with the map visible:
![image](https://github.com/RPTools/maptool/assets/7492219/0e6e1e80-08d4-4ac6-9439-6d155b90dd9f)

### MapTool Info

1.13.2

### Desktop

Linux Mint 21.2

### Additional Context

On 1.13.2, this issue seems to be 100% consistent. I.e., the impersonation is never cleared.

Tested on 1.14.0-rc.4 and latest develop as well, where the situation is more subtle. The first time the map is hidden, the impersonation window is cleared, though the chat window still shows the impersonation as active. If the map is made visible and then hidden again right away, then the impersonation window is not cleared. This seems to be related to the fact that there is a selection change the first time the map is hidden, but there is none the second time.

**Comment by cwisniew** [comment]
I wonder if we should make map visibility changes an event and have all interested parties listen for that event, rather than depending on other mechanisms for clearing out things when a map is made non visible

**Comment by kwvanderlinde** [comment]
Hiding a map necessarily clears out the current zone for player clients on that map, so they should see a `ZoneDeactivated` event. But currently it doesn't fire as we don't switch them to a different zone.

But that's easily fixed. And then all components can use `ZoneActivated` / `ZoneDeactivated` to stay informed about the current zone and update accordingly. That would handle map visibility changes, as well as any other means of a zone being changed.

**Comment by cwisniew** [comment]
> Hiding a map necessarily clears out the current zone for player clients on that map, so they should see a `ZoneDeactivated` event. But currently it doesn't fire as we don't switch them to a different zone.
> 
> But that's easily fixed. And then all components can use `ZoneActivated` / `ZoneDeactivated` to stay informed about the current zone and update accordingly. That would handle map visibility changes, as well as any other means of a zone being changed.

This is only the case when the zone being hidden is the current active zone, which works for fixing this defect but does not offer the functionality provided by a zone visibility change event which would be useful elsewhere in the future.

**Comment by kwvanderlinde** [comment]
Yeah, but I believe that is all we need. In terms of interacting with zones, we have this kind of pattern:
1. Some things (certain UI elements) are attached to the current zone and so only need to stay up-to-date with the changes to the current zone (including whether there is a current zone).
2. Other functionality (e.g., macro functions, plus some UI like the map selector) look up zones on demand, whether the current zone or otherwise.

(1) can served by existing events, and (2) wouldn't really benefit from events. I'm wondering what might be coming that wouldn't/shouldn't fit into those cases, which would need to react to changes in some non-current zone.

**Comment by cwisniew** [comment]
> 
> (1) can served by existing events, and (2) wouldn't really benefit from events. I'm wondering what might be coming that wouldn't/shouldn't fit into those cases, which would need to react to changes in some non-current zone.

On my -- admittedly too long -- list of things to do is an updated map selector that can be docked to the side if required, which would better support things like grouping zones into a folder-like structure, GM hiding/unhiding with fewer clicks etc., which would require maintaining a list of zones that are available and visibility.

**Comment by kwvanderlinde** [comment]
I figured that might be it :smile: 

... in which case a more general event might be useful. More of a "this map was modified _in some way_, now update yourself". Otherwise you'll also want a "map name change", "map display name change", "map grouping changed", etc., depending on just how many properties end up informing that view.

**Comment by cwisniew** [comment]
> I figured that might be it 😄
> 
> ... in which case a more general event might be useful. More of a "this map was modified _in some way_, now update yourself". Otherwise you'll also want a "map name change", "map display name change", "map grouping changed", etc., depending on just how many properties end up informing that view.

I guess I can live with the change event :)
